### PR TITLE
Index controller

### PR DESCRIPTION
**Pull request by jenow**
## What does this PR do ?

depends on #13 

Add the documentation for index controller
Also fix a minor issue in index:mDelete

### CODE REVIEW

File: doc/2/controllers/document/count/index.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+---
+
+## Arguments
+
+```dart
+Future<int> count(
+    String index,
+    String collection,
+    {Map<String, dynamic> query})
```
Comment (y-abs) [inline doc/2/controllers/document/count/index.md]:
```suggestion
    { Map<String, dynamic> query, })
```
:trollface:

File: doc/2/controllers/document/count/index.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+| ------------------ | -------------------------------------------- | --------------- |
+| `index`            | <pre>String</pre>                            | Index name      |
+| `collection`       | <pre>String</pre>                            | Collection name |
+| `query`      | <pre>Map<String, dynamic></pre><br>(`{}`) | Query to match  |
+
+---
+
+## Return
+
+Returns an Integer.
```
Comment (y-abs) [inline doc/2/controllers/document/count/index.md]:
```suggestion
Returns an int.
```

File: doc/2/controllers/document/create/snippets/create.dart [hunk 3]
```diff
@@ -0,0 +1,10 @@
+
+final result = await kuzzle
+  .document
+  .create(
+    'nyc-open-data',
+    'yellow-taxi',
+    {
+      'firstname': 'John'
```
Comment (y-abs) [inline doc/2/controllers/document/create/snippets/create.dart]:
```suggestion
      'firstname': 'John',
```

File: doc/2/controllers/document/delete-by-query/snippets/delete-by-query.dart [hunk 4]
```diff
@@ -0,0 +1,9 @@
+final result = await kuzzle
+  .document
+  .deleteByQuery('nyc-open-data', 'yellow-taxi', {
+    'query': {
+      'match': {
+        'capacity': 4
```
Comment (y-abs) [inline doc/2/controllers/document/delete-by-query/snippets/delete-by-query.dart]:
```suggestion
        'capacity': 4,
```

File: doc/2/controllers/document/m-create-or-replace/index.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    List<Map<String, dynamic>> documents, {
+    bool waitForRefresh = false,
+  })
+```
+
+| Arguments          | Type                                                    | Description                       |
+| ------------------ | ------------------------------------------------------- | --------------------------------- |
+| `index`            | <pre>String</pre>                                       | Index                             |
+| `collection`       | <pre>String</pre>                                       | Collection                        |
+| `documents`        | <pre>List<Map<String, dynamic>></pre> | ArrayList containing the documents to create |
```
Comment (y-abs) [inline doc/2/controllers/document/m-create-or-replace/index.md]:
```suggestion
| `documents`        | <pre>List<Map<String, dynamic>></pre> | List containing the documents to create |
```

File: doc/2/controllers/document/m-create/index.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+---
+
+## Arguments
+
+```dart
+Future<c> mCreate(
+      String index, String collection, 
+      List<Map<String, dynamic>> documents,
+      {bool waitForRefresh = false})
```
Comment (y-abs) [inline doc/2/controllers/document/m-create/index.md]:
```suggestion
      { bool waitForRefresh = false, })
```

File: doc/2/controllers/document/m-create/index.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+```
+
+| Arguments          | Type                                                    | Description                       |
+| ------------------ | ------------------------------------------------------- | --------------------------------- |
+| `index`            | <pre>String</pre>                                       | Index                             |
+| `collection`       | <pre>String</pre>                                       | Collection                        |
+| `documents`        | <pre>List<Map<String, dynamic>></pre> | ArrayList containing the documents to create |
```
Comment (y-abs) [inline doc/2/controllers/document/m-create/index.md]:
```suggestion
| `documents`        | <pre>List<Map<String, dynamic>></pre> | List containing the documents to create |
```

File: doc/2/controllers/document/m-create/index.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+---
+
+### documents
+
+Each document has the following properties:
+
+| Arguments          | Type                                         | Description                       |
+| ------------------ | -------------------------------------------- | --------------------------------- |
+| `_id`              | <pre>String</pre>                            | Optional document ID. Will be auto-generated if not defined.             |
+| `body`             | <pre>List<Map<String, dynamic>></pre> | Document body |
```
Comment (y-abs) [inline doc/2/controllers/document/m-create/index.md]:
```suggestion
| `body`             | <pre>Map<String, dynamic></pre> | Document body |
```

File: doc/2/controllers/document/m-create/index.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+## Return
+
+A `List<Map<String, dynamic>>` which has a `successes` and `errors`:
```
Comment (y-abs) [inline doc/2/controllers/document/m-create/index.md]:
```suggestion
A `Map<String, dynamic>` which has a `successes` and `errors`:
```

File: doc/2/controllers/document/m-create/index.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+Each created document is an object of the `successes` array with the following properties:
+
+| Property     | Type                                         | Description                      |
+|------------- |--------------------------------------------- |--------------------------------- |
+| `_source`    | <pre>List<Map<String, dynamic>></pre> | Created document                 |
```
Comment (y-abs) [inline doc/2/controllers/document/m-create/index.md]:
```suggestion
| `_source`    | <pre>Map<String, dynamic></pre> | Created document                 |
```

File: doc/2/controllers/document/m-create/index.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+| `_id`        | <pre>String</pre>                            | ID of the newly created document                       |
+| `_version`   | <pre>int</pre>                           | Version of the document in the persistent data storage |
+
+Each errored document is an object of the `errors` array with the following properties:
+
+| Property     | Type                                         | Description                      |
+|------------- |--------------------------------------------- |--------------------------------- |
+| `document`   | <pre>List<Map<String, dynamic>></pre> | Document that causes the error   |
```
Comment (y-abs) [inline doc/2/controllers/document/m-create/index.md]:
```suggestion
| `document`   | <pre>Map<String, dynamic></pre> | Document that causes the error   |
```

File: doc/2/controllers/document/m-create/snippets/m-create.dart [hunk 12]
```diff
@@ -0,0 +1,16 @@
+final result = await kuzzle
+  .document
+  .mCreate("nyc-open-data", "yellow-taxi", [
+    {
+      '_id': 'some-id',
+      'body': {
+        'Agent': 'Smith'
```
Comment (y-abs) [inline doc/2/controllers/document/m-create/snippets/m-create.dart]:
```suggestion
        'Agent': 'Smith',
```

File: doc/2/controllers/document/m-create/snippets/m-create.dart [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+      }
+    },
+    {
+      '_id': 'some-id2',
+      'body': {
+        'Gordon': 'Freeman'
```
Comment (y-abs) [inline doc/2/controllers/document/m-create/snippets/m-create.dart]:
```suggestion
        'Gordon': 'Freeman',
```

File: doc/2/controllers/document/m-delete/snippets/m-delete.dart [hunk 14]
```diff
@@ -0,0 +1,6 @@
+final result = await kuzzle
+  .document
+  .mDelete('nyc-open-data', 'yellow-taxi', [
+    'some-id',
+    'some-id2'
```
Comment (y-abs) [inline doc/2/controllers/document/m-delete/snippets/m-delete.dart]:
```suggestion
    'some-id2',
```

File: doc/2/controllers/document/m-get/index.md [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+| Arguments          | Type                                                    | Description                       |
+| ------------------ | ------------------------------------------------------- | --------------------------------- |
+| `index`            | <pre>String</pre>                                       | Index name                        |
+| `collection`       | <pre>String</pre>                                       | Collection name                   |
+| `ids`              | <pre>List<String></pre>                            | Document IDs                      |
+---
+
+## Return
+
+A `List<String>` which has a `successes` and `errors`:
```
Comment (y-abs) [inline doc/2/controllers/document/m-get/index.md]:
```suggestion
A `Map<String, dynamic>` which has a `successes` and `errors`:
```

File: doc/2/controllers/document/m-get/index.md [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+Each created document is an object of the `successes` array with the following properties:
+
+| Property     | Type                                         | Description                      |
+|------------- |--------------------------------------------- |--------------------------------- |
+| `_source`    | <pre>List<String></pre> | Document content                 |
```
Comment (y-abs) [inline doc/2/controllers/document/m-get/index.md]:
```suggestion
| `_source`    | <pre>Map<String, dynamic></pre> | Document content                 |
```

File: doc/2/controllers/document/m-get/index.md [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+| `_id`        | <pre>String</pre>                            | Document ID                      |
+| `_version`   | <pre>int</pre>                           | Version of the document in the persistent data storage |
+
+The `errors` array contain the IDs of not found documents.
```
Comment (y-abs) [inline doc/2/controllers/document/m-get/index.md]:
```suggestion
The `errors` array contains the IDs of not found documents.
```

File: doc/2/controllers/document/m-replace/index.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    List<Map<String, dynamic>> documents, {
+    bool waitForRefresh = false,
+  })
+```
+
+| Arguments          | Type                                                    | Description                       |
+| ------------------ | ------------------------------------------------------- | --------------------------------- |
+| `index`            | <pre>String</pre>                                       | Index                             |
+| `collection`       | <pre>String</pre>                                       | Collection                        |
+| `documents`        | <pre>Map<String, dynamic></pre> | ArrayList containing the documents to replace |
```
Comment (y-abs) [inline doc/2/controllers/document/m-replace/index.md]:
```suggestion
| `documents`        | <pre>List<Map<String, dynamic>></pre> | List containing the documents to replace |
```

File: doc/2/controllers/document/m-update/index.md [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+---
+
+## Arguments
+
+```dart
+Future<Map<String, dynamic>> mUpdate(
+      String index, String collection, 
+      List<Map<String, dynamic>> documents,
+      {bool waitForRefresh = false, int retryOnConflict})
```
Comment (y-abs) [inline doc/2/controllers/document/m-update/index.md]:
```suggestion
      { bool waitForRefresh = false, int retryOnConflict, })
```

File: doc/2/controllers/document/m-update/index.md [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+```
+
+| Arguments          | Type                                                    | Description                       |
+| ------------------ | ------------------------------------------------------- | --------------------------------- |
+| `index`            | <pre>String</pre>                                       | Index                             |
+| `collection`       | <pre>String</pre>                                       | Collection                        |
+| `documents`        | <pre>List<Map<String, dynamic>></pre> | ArrayList containing the documents to update |
```
Comment (y-abs) [inline doc/2/controllers/document/m-update/index.md]:
```suggestion
| `documents`        | <pre>List<Map<String, dynamic>></pre> | List containing the documents to update |
```

File: doc/2/controllers/document/m-update/index.md [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+Each document has the following properties:
+
+| Arguments          | Type                                         | Description                       |
+| ------------------ | -------------------------------------------- | --------------------------------- |
+| `_id`              | <pre>String</pre>                            | Document ID             |
+| `body`             | <pre>Map<String, dynamic></pre> | Document body |
+
+## Return
+
+A `ConcurrentHashMap<String, ArrayList<Object>>` which has a `successes` and `errors`:
```
Comment (y-abs) [inline doc/2/controllers/document/m-update/index.md]:
```suggestion
A `Map<String, dynamic>` which has a `successes` and `errors`:
```

File: doc/2/controllers/document/replace/index.md [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+| `collection`       | <pre>String</pre>                            | Collection                        |
+| `id`               | <pre>String</pre>                            | Document ID                       |
+| `document`         | <pre>Map<String, dynamic></pre> | New content of the document to update |
+| `waitForRefresh`   | <pre>bool</pre><br>(`false`)                 | If set to `true`, Kuzzle will wait for the persistence layer to finish indexing|
+
+---
+
+## Return
+
+A `ConcurrentHashMap` which has the following properties:
```
Comment (y-abs) [inline doc/2/controllers/document/replace/index.md]:
```suggestion
A `Map` which has the following properties:
```

File: doc/2/controllers/document/search/index.md [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+    String scroll,
+  })
+```
+ 
+| Arguments          | Type                                         | Description                       |
+| ------------------ | -------------------------------------------- | --------------------------------- |
+| `index`            | <pre>String</pre>                            | Index                             |
+| `collection`       | <pre>String</pre>                            | Collection                        |
+| `searchQuery`      | <pre>Map<String, dynamic></pre>                 | Search query                   |
+| `from`     | <pre>Integer</pre><br/>(`0`)    | Offset of the first document to fetch                  |
```
Comment (y-abs) [inline doc/2/controllers/document/search/index.md]:
```suggestion
| `from`     | <pre>int</pre><br/>(`0`)    | Offset of the first document to fetch                  |
```

File: doc/2/controllers/document/search/index.md [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+| `size`     | <pre>Integer</pre><br/>(`10`)   | Maximum number of documents to retrieve per page       |
```
Comment (y-abs) [inline doc/2/controllers/document/search/index.md]:
```suggestion
| `size`     | <pre>int</pre><br/>(`10`)   | Maximum number of documents to retrieve per page       |
```

File: doc/2/controllers/realtime/publish/index.md [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+```dart
+Future<bool> publish(
+      String index, String collection, Map<String, dynamic> message)
+```
+
+| Argument     | Type               | Description                         |
+|--------------|--------------------|-------------------------------------|
+| `index`      | <pre>String</pre>  | Index name                          |
+| `collection` | <pre>String</pre>  | Collection name                     |
+| `message`    | <pre>Map<String, dynamic></pre> | ConcurrentHashMap representing a JSON payload |
```
Comment (y-abs) [inline doc/2/controllers/realtime/publish/index.md]:
```suggestion
| `message`    | <pre>Map<String, dynamic></pre> | Map representing a JSON payload |
```

**Review by y-abs — CHANGES_REQUESTED** [review]

**Review by y-abs — APPROVED** [review]

**Review by alexandrebouthinon — APPROVED** [review]

**Pull request merged**